### PR TITLE
Improvement for EshelbyTensor calculation

### LINKED ISSUE #10222: Improve EshelbyTensor calculation
## Rationale

Two issues with existing EshelbyTensor class. 1. It uses incremental strain material property only. However, for small strain calculations, total strain is calculated but incremental strain is not. Thus the EshelbyTensor does not work with small strain case. 2. There are two equivalent way to calculate the EshelbyTensor, one is to use the deformation gradient F and the other is to use the gradient of displacements, H. For the small strain case, the H gives much better results than F. For other cases, they are very close. Thus, to use H seems to be a better option.  

## Description

Added the total strain option and replace the F with H. Need to regold 16 tests files related to J-integral calculation in tensor_mechanics module. 

## Impact

Only affect fracture J-intergral staff.

### PR DESCRIPTION
**Pull request by jiangwen84**
Added total strain option
Replaced F with H
Regold tests
Added one regression test for the small strain formulation which uses total_strain (actually mechanical_strain)

closes #10222

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/EshelbyTensor.C [hunk 1]
```diff
@@ -73,21 +86,26 @@ EshelbyTensor::initQpStatefulProperties()
...
+  else
+    _sed[_qp] = _stress[_qp].doubleContraction((*_mechanical_strain)[_qp]) / 2.0;
+
   // 1st Piola-Kirchoff Stress (P):
   RankTwoTensor P = detF * _stress[_qp] * FinvT;
 
-  // FTP = F^T * P = F^T * detF * sigma * FinvT;
-  RankTwoTensor FTP = F.transpose() * P;
+  // FTP = H^T * P = H^T * detF * sigma * FinvT;
+  RankTwoTensor FTP = H.transpose() * P;
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/EshelbyTensor.C]: Name this HTP

**Comment by jiangwen84** [comment]
@bwspenc  Can you further review this PR? Thanks.

File: modules/solid_mechanics/test/tests/domain_integral_thermal/gold/j_integral_2d_out.e [hunk 2]
```diff
@@ -1 +0,0 @@
-../../../../../tensor_mechanics/test/tests/domain_integral_thermal/gold/j_integral_2d_out.e
```
Comment (bwspenc) [inline modules/solid_mechanics/test/tests/domain_integral_thermal/gold/j_integral_2d_out.e]: It looks like these were all symlinks, and you replaced them with new gold files.  We should be able to still symlink to the tensor_mechanics results to make sure they produce the same results, shouldn't we?

File: modules/tensor_mechanics/src/materials/EshelbyTensor.C [hunk 3]
```diff
@@ -62,6 +61,20 @@ EshelbyTensor::EshelbyTensor(const InputParameters & parameters)
...
+    _strain_increment = &getMaterialProperty<RankTwoTensor>(_base_name + "strain_increment");
+    _mechanical_strain = nullptr;
+  }
+  else if (hasMaterialProperty<RankTwoTensor>(_base_name + "mechanical_strain"))
+  {
+    _mechanical_strain = &getMaterialProperty<RankTwoTensor>(_base_name + "mechanical_strain");
+    _strain_increment = nullptr;
+  }
+  else
+    mooseError("EshelbyTensor cannot find eigher mechanical_strain or strain_increment material "
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/EshelbyTensor.C]: eigher->either

**Review by bwspenc — CHANGES_REQUESTED** [review]
The code changes in tensor_mechanics look good.  I'd prefer that you make the same changes in solid_mechanics and keep the symlinks to the results files.

Also, there's a documentation page on this in:
moose/docs/content/documentation/systems/Materials/tensor_mechanics/EshelbyTensor.md that needs a minor update to reflect these changes.

**Comment by jiangwen84** [comment]
@bwspenc all the comments are addressed :-)

**Review by bwspenc — APPROVED** [review]

**Pull request merged**